File: observer/observer_test.go
package observer_test

import (
	"design_pattern/observer"
	"testing"
)

func TestObserver(t *testing.T) {
	t.Run("Observer run", func(t *testing.T) {
		messenger := observer.NewMessenger()
		mailSender := observer.NewMailSender()
		appAlarmSender := observer.NewAppAlarmSender()

		messenger.Subscribe(mailSender)
		messenger.Subscribe(appAlarmSender)

		messenger.Publish("Show me the Money!")

		messenger.Unsubscribe(appAlarmSender)

		messenger.Publish("Give me the Money!")
	})
}

File: observer/observer.test.ts
import { Messenger } from "./messenger";
import { AppAlarmSender } from "./sender/app-alarm-sender";
import { MailSender } from "./sender/mail-sender";

test('Observer Run', () => {
    const messenger = new Messenger();
    const mailSender = new MailSender();
    const appAlarmSender = new AppAlarmSender();

    messenger.subscribe(mailSender);
    messenger.subscribe(appAlarmSender);

    messenger.publish("Show me the Money!");

    messenger.unsubscribe(appAlarmSender);

    messenger.publish("Give me the Money!");
});